### DISCUSSION TITLE
Routing problem for a single driver

### DISCUSSION THREAD

**Discussion by ibiscp** [google/or-tools#4213] [Routing (and legacy CP) questions] [answered]
Hello everyone,

I am currently exploring different solutions for a scheduling problem and need some guidance. The challenge involves planning a month's worth of visits to various locations for a single driver, each with specific time windows.

The schedule needs to be organized into shifts, and additionally, some locations require multiple visits with a certain time gap between each. I haven't found a scenario that matches these exact requirements. Could anyone suggest how I might approach this, or point me towards any resources or tools that could help?

Thank you in advance for your assistance!

**Comment by sschnug** [comment, marked as answer]
Multi-visit implies a need for node-duplication (`visit_x -> {visit_x_0, visit_x_1, ...}`).

"certain time-gap" is just a constraint on the time-accumulation-dimension similar to time-windows or pickup-and-delivery ([Link](https://developers.google.com/optimization/routing/pickup_delivery#python)), e.g.

        MIN_GAP = 100
        routing.solver().Add(
            distance_dimension.CumulVar(visit_x_0) + MIN_GAP
            <= distance_dimension.CumulVar(visit_x_1)
        )


This in conjunction with all the examples in the repository should be enough to do some experiments.

Of course a lot of details are missing, but the task itself is prescribed too informally to tackle those. The main hurdle might be whatever "shift" means to you. It might make sense to not only duplicate visit-nodes (multi-visit) but also vehicles/drivers (while constraining their available-time) to emulate shifts.
